include/nbl/builtin/hlsl/cpp_compat/matrix.h

```diff
@@ -22,42 +22,35 @@ struct matrix final : private glm::mat<N,M,T>
     using Base::operator=;
     using Base::operator[];
 
-    template<uint16_t X, uint16_t Y> matrix(matrix<T, X, Y> const&) = delete;
+    template<uint16_t X, uint16_t Y, std::enable_if<!(X == N && Y == M) && X <= N && Y <= M>>
+    matrix(matrix<T, X, Y> const& m) : Base(m)
+    {
+    }
 
     matrix(matrix const&) = default;
     explicit matrix(Base const& base) : Base(base) {}
```

Comment:
having both this and `using Base::Base` makes no sense

Btw won't `using Base::Base` and `Base::operator=` give you public constructors and assignment operators that accept `Base` and not `matrix`? (A definite DO NOT WANT)

Reply:
Well if we dont use Base::Base we would have to write out every single constructor manually. And by declaring the matrix constructor that takes Base we force it to be explicit. However, instead of Base::operator= we could default the assignment operator probably.

Reply:
ok do that, make sure we can never implicitly convert `matrix` to `Base`, neither when assigning or passing as function arg or in some operator.